File: center/receiver/radio.py

from . import cc1101

class Radio(cc1101.CC1101):
    async def calibrate(self):
        self.wcmd(self.CommandStrobe.SCAL)
        await self.waitstate(1)

    async def setup_basic(self):
        await self.reset()

        # frequency configuration
        self.set(self.ConfReg.FREQ2, 0x10)
        self.set(self.ConfReg.FREQ1, 0xa7)
        self.set(self.ConfReg.FREQ0, 0xe1)

        # modem configuration
        self.set(self.ConfReg.MDMCFG4, 0xaa)
        self.set(self.ConfReg.MDMCFG3, 0xbc)
        self.set(self.ConfReg.MDMCFG2, 0x71)
        self.set(self.ConfReg.DEVIATN, 0x40)

        # PATABLE
        self.set(self.ConfReg.PATABLE, 0xc0)

    async def setup_for_rx(self):
        # fix packet length
        self.set(self.ConfReg.PKTLEN, 16)

        # packet automation
        self.set(self.ConfReg.PKTCTRL0, 0x44)

        self.set(self.ConfReg.PKTCTRL1, 0x2c)

        # main radio control state machine configuration
        self.set(self.ConfReg.MCSM1, 0x3c)
        self.set(self.ConfReg.MCSM0, 0x34)

        self.set(self.ConfReg.IOCFG0, 0x07)

        await self.calibrate()

    async def setup_for_conf(self):
        # treat 0 as broadcast
        self.set(self.ConfReg.PKTLEN, 0x10)
        self.set(self.ConfReg.PKTCTRL1, 0x0a)
        self.set(self.ConfReg.PKTCTRL0, 0x45)
        self.set(self.ConfReg.IOCFG0, 0x07)
        self.set(self.ConfReg.MCSM1, 0x0b)
        self.set(self.ConfReg.ADDR, 0)

        await self.calibrate()

File: receiver/radio.go
package receiver

import (
	"github.com/rkojedzinszky/thermo-center/v5/receiver/cc1101"
)

type radio struct {
	cc1101 *cc1101.CC1101
}

func (r radio) calibrate() (err error) {
	if err = r.cc1101.Cmd(cc1101.SCAL); err != nil {
		return
	}

	return r.cc1101.Waitstate(1)
}

type configSetting struct {
	confreg cc1101.ConfReg
	value   uint8
}

// Basic radio configuration
var basicSettings = []configSetting{
	// Frequency
	{cc1101.FREQ2, 0x10},
	{cc1101.FREQ1, 0xa7},
	{cc1101.FREQ0, 0xe1},

	// Modem
	{cc1101.MDMCFG4, 0xaa},
	{cc1101.MDMCFG3, 0xbc},
	{cc1101.MDMCFG2, 0x71},
	{cc1101.DEVIATN, 0x40},

	// PATABLE
	{cc1101.PATABLE, 0xc0},
}

var rxSettings = []configSetting{
	// fix packet length
	{cc1101.PKTLEN, 16},

	// packet automation
	{cc1101.PKTCTRL1, 0x2c},
	{cc1101.PKTCTRL0, 0x44},

	// main radio control state machine configuration
	{cc1101.MCSM1, 0x3c},
	{cc1101.MCSM0, 0x34},

	// Interrupt on receive
	{cc1101.IOCFG0, 0x07},
}

var confSettings = []configSetting{
	// fix packet length
	{cc1101.PKTLEN, 16},

	// packet automation, treat 0 as broadcast
	{cc1101.PKTCTRL1, 0x0a},
	{cc1101.PKTCTRL0, 0x45},
	{cc1101.ADDR, 0},

	{cc1101.MCSM1, 0x08},

	// Interrupt on receive
	{cc1101.IOCFG0, 0x07},
}

func (r radio) applySettings(settings []configSetting) error {
	for _, s := range settings {
		if err := r.cc1101.SetConf(s.confreg, s.value); err != nil {
			return err
		}
	}

	return nil
}

func (r radio) setupBasic() error {
	return r.applySettings(basicSettings)
}

func (r radio) setupForRX() error {
	if err := r.applySettings(rxSettings); err != nil {
		return err
	}

	if err := r.calibrate(); err != nil {
		return err
	}

	return r.cc1101.Cmd(cc1101.SRX)
}

func (r radio) setupForConf() error {
	if err := r.applySettings(confSettings); err != nil {
		return err
	}

	return r.calibrate()
}
